nixos/modules/services/hardware/evdevremapkeys.nix

```diff
@@ -0,0 +1,79 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.evdevremapkeys;
+
+  fullConfig = {
+    devices = cfg.devices;
+  };
+
+  cfgFile = pkgs.writeText "config.yaml" (builtins.toJSON fullConfig);
+in {
+
+  options.services.evdevremapkeys = {
+
+    # Note: this corresponds to the YAML configuration that evdevremap keys
+    # expects.
+    devices = mkOption {
+      default = {};
+      description = "Input devices to remap using evdevremapkeys";
+      type = with types; listOf (submodule {
+        options = {
+
+          input_name = mkOption {
+            type = str;
+            description = ''
+              Name of device to be remapped (from `xinput list`).
+            '';
+          };
+
+          output_name = mkOption {
+            type = str;
+            description = ''
+              Name of virtual device that will output remapped events.
+            '';
+          };
+
+          remappings = mkOption {
+            type = attrsOf (listOf str);
+            description = ''
+              Attrset of inputted event name to list of events that will be
+              fired instead. The event names are standard Linux input event
+              names, for a list see include/uapi/linux/input-event-codes.h in
+              the Linux source tree.
+
+              For example:
+                remappings = {
+                  KEY_ESC = [
+                    "KEY_A"
+                    "KEY_B"
+                  ];
+                };
+              will remap the escape key to instead fire an A and B keypress
+              simultaneously when pressed.
+            '';
+          };
+        };
+      });
+    };
+
+  };
+
+  config = mkIf (cfg.devices != []) {
+
+    systemd.services.evdevremapkeys = {
+      description = "Remap Evdev Input";
+      after = [ "systemd-udev-settle.service" ];
```

Comment:
I was reading the manpage of `systemd-udev-settle` and it said:

        Using this service is not recommended.  There can be no guarantee that hardware is fully discovered at any specific
       time, because the kernel does hardware detection asynchronously, and certain buses and devices take a very long
       time to become ready, and also additional hardware may be plugged in at any time. Instead, services should
       subscribe to udev events and react to any new hardware as it is discovered. Services that, based on configuration,
       expect certain devices to appear, may warn or report failure after a timeout. This timeout should be tailored to
       the hardware type. Waiting for systemd-udev-settle.service usually slows boot significantly, because it means
       waiting for all unrelated events too.


I also saw that the example upstream provides doesn't depend on it https://github.com/philipl/evdevremapkeys/blob/master/examples/evdevremapkeys.service

Why do we need it?


Also; wouldn't this service make more sense to run as a user unit instead of a system unit?

Reply:
From the source code of `evdevremapkeys` it seems they subscribe to `udev` events; so I don't think there is any need to wait for `systemd-udev-settle`

https://github.com/philipl/evdevremapkeys/blob/6626275333818fbdedc9b6ab20dc12531e29b399/evdevremapkeys/evdevremapkeys.py#L344-L345

Reply:
Whoops, I did indeed copy this mindlessly this from somewhere else in nixpkgs during development and then forgot to actually make this sensible :). I've updated the service definition to reflect upstream's service example.